hail/src/main/scala/is/hail/annotations/CodeOrdering.scala

```diff
@@ -68,294 +57,518 @@ object CodeOrdering {
         if (sortOrders == null) Ascending else sortOrders(i),
         op)
 
-    override def compareNonnull(x: Code[Long], y: Code[Long]): Code[Int] = {
-      val cmp = mb.newLocal[Int]()
+    override def compareNonnull(cb: EmitCodeBuilder, x: PCode, y: PCode): Code[Int] = {
```

Comment:
why do we need to implement all the orderings? Can't we implement them all in terms of `compareNonNull`?

Reply:
Yes. It is correct to do so, but if the comparison is significantly more expensive than `lt` or `eq` then we end up doing extra work.